dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/config/MetadataUtils.java

```diff
@@ -0,0 +1,126 @@
+
+    public static final TraceComponent tc = Tr.register(MetadataUtils.class);
+
+    private static final String KEY_EP_REQUEST = "endpointRequest";
+    private static volatile EndpointRequest endpointRequest;
+
+    @Reference(name = KEY_EP_REQUEST, policy = ReferencePolicy.DYNAMIC)
+    public void setEndpointRequest(EndpointRequest endpointRequestService) {
+        endpointRequest = endpointRequestService;
+    }
+
+    public void unsetEndpointRequest(EndpointRequest endpointRequestService) {
+        endpointRequest = null;
+    }
+
+    /**
+     * Returns a parameterized value from the configured OidcProviderMetadata, or from the OP's discovery document if the value
+     * cannot be found in the OidcProviderMetadata.
+     *
+     * @throws OidcDiscoveryException Thrown if the value cannot be found in the discovery document or if its value is empty.
+     * @throws OidcClientConfigurationException Thrown if the client configuration is missing the providerURI.
+     */
+    public static <T> T getValueFromProviderOrDiscoveryMetadata(OidcClientConfig oidcClientConfig, Function<OidcProviderMetadata, T> metadataMethodToCall,
+                                                                String discoveryMetadataKey) throws OidcDiscoveryException, OidcClientConfigurationException {
+        OidcProviderMetadata providerMetadata = oidcClientConfig.getProviderMetadata();
+        if (providerMetadata != null) {
+            T value = metadataMethodToCall.apply(providerMetadata);
+            if (value != null && !value.toString().isEmpty()) {
+                if (tc.isDebugEnabled()) {
+                    Tr.debug(tc, discoveryMetadataKey + " found in the provider metadata: [" + value + "]");
+                }
+                return value;
+            }
+        }
+        return getValueFromDiscoveryMetadata(oidcClientConfig, discoveryMetadataKey);
+    }
+
+    /**
+     * Returns a parameterized value from the OP's discovery document.
+     *
+     * @throws OidcDiscoveryException Thrown if the value cannot be found in the discovery document or if its value is empty.
+     * @throws OidcClientConfigurationException Thrown if the client configuration is missing the providerURI.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> T getValueFromDiscoveryMetadata(OidcClientConfig oidcClientConfig, String key) throws OidcDiscoveryException, OidcClientConfigurationException {
+        T value = null;
+        JSONObject providerDiscoveryMetadata = endpointRequest.getProviderDiscoveryMetadata(oidcClientConfig);
+        if (providerDiscoveryMetadata != null) {
+            value = (T) providerDiscoveryMetadata.get(key);
+        }
+        if (value == null || value.toString().isEmpty()) {
+            String nlsMessage = Tr.formatMessage(tc, "DISCOVERY_METADATA_MISSING_VALUE", key);
+            throw new OidcDiscoveryException(oidcClientConfig.getClientId(), oidcClientConfig.getProviderURI(), nlsMessage);
+        }
```

Comment:
i think we should add a check to see if the attribute is required or not before throwing the exception in the cases of `userinfo_endpoint` and `end_session_endpoint`

Reply:
I'd like to keep the exception being thrown in this method and allow the caller to decide what to do with it depending on the situation. For example, failing to find the UserInfo endpoint will catch the exception, ignore the FFDC, log a debug message saying UserInfo claims couldn't be built, and then move on harmlessly. On the other hand, failing to get the authorization endpoint will catch the exception, log an error message, and fail the request.

You're right in that there are some paths that don't handle the exception as they should with my changes. Looks like I'll need to fix how the following calls are handled:
- `getEndSessionEndpoint()`: The logout code should be able to function gracefully if the end session endpoint can't be found.
- `getJwksUri()`: The ID token validation code shouldn't immediately fail if a JWKS URI can't be found since the keys can be found elsewhere.

Reply:
for the jwksURI - i thought we decided last week to fail when there isn't a jwksURI in the metadata and not use fallback methods?

(refer: https://github.com/OpenLiberty/open-liberty/pull/22990)

Reply:
I agree Adam about letting the caller decide how to handle the exception. I agree Jimmy about fixing those paths to handle the exception correctly.

Reply:
Default trust store can also contain signing key as it is the case with openidConnectClient, but spec requires jwks URI. It must be available from the discovery document or provided in the annotation.

Reply:
yes, let's discuss this later -- approving for now

Reply:
True, I forgot we did talk about failing if the JWKS URI can't be found since the spec requires it. I'd like to have us fail for now if we can't find it, and looking elsewhere for the keys can be a value-add we can consider later. I'll work on another commit to do it in this PR.